Scalafix is not able to parse Scala 3 nightly version

scalafix fails to parse nightly version of Scala3 which differs from nightly version of Scala 3
Example of version `3.6.0-RC1-bin-20241008-3408ed7-NIGHTLY` 

Issue found by Scala 3 Open Community Build in 3 projects using automatic application of scalafix on compile: 
* gemini-hlsw/lucuma-schemas -	[Open CB logs](https://github.com/VirtusLab/community-build3/actions/runs/11531682951/job/32102782702)	
* tkrs/orcus - [Open CB logs](https://github.com/VirtusLab/community-build3/actions/runs/11531682451/job/32102750477)	
* xuwei-k/unused-proto	- [Open CB logs](https://github.com/VirtusLab/community-build3/actions/runs/11531682951/job/32102794393)

Hello, any news on that front? `Scala 3.6.2` is out and `0.13.0` can't parse it either e.g. if fails on new multiple bounds syntax (the have to be enclosed with `{}`) with:
```
[error] .../Foo.scala:151:34: error: illegal start of declaration
[error]   private def bar[F[_]: { MonadThrow, Concurrent }](
[error]                                  ^^^^^^^^^^

```

> Hello, any news on that front? `Scala 3.6.2` is out and `0.13.0` can't parse it either e.g. it fails on new multiple bounds syntax (the have to be enclosed with `{}`) with:
> 
> ```
> [error] .../Foo.scala:151:34: error: illegal start of declaration
> [error]   private def bar[F[_]: { MonadThrow, Concurrent }](
> [error]                                  ^^^^^^^^^^
> ```

It's not related to this issue, you might be intrested in https://github.com/scalameta/scalameta/pull/4068 that introduced new givens syntax to scalameta parser used by scalafix. 

This issue was about not being able to version strings used by Scala 3 nightlies, not the Scala 3 syntax

--- a/scalafix-core/src/main/scala/scalafix/internal/config/ScalaVersion.scala
+++ b/scalafix-core/src/main/scala/scalafix/internal/config/ScalaVersion.scala
@@ -1,5 +1,8 @@
 package scalafix.internal.config
 
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+
 import scala.util.Failure
 import scala.util.Success
 import scala.util.Try
@@ -44,9 +47,21 @@ sealed trait ScalaVersion {
     case Minor(major, minorVersion) => s"${major.value}.${minorVersion}"
     case Patch(major, minorVersion, patchVersion) =>
       s"${major.value}.${minorVersion}.$patchVersion"
-    case RC(major, minorVersion, patchVersion, rc) =>
-      s"${major.value}.${minorVersion}.$patchVersion-RC$rc"
-    case Nightly(major, minorVersion, patchVersion, shortSha1) =>
+    case RC(major, minorVersion, patchVersion, rcVersion) =>
+      s"${major.value}.${minorVersion}.$patchVersion-RC$rcVersion"
+    // Scala 3 nightly
+    case Nightly(
+          major,
+          minorVersion,
+          patchVersion,
+          maybeRcVersion,
+          Some(dateVersion),
+          shortSha1
+        ) =>
+      val maybeRcSuffix = maybeRcVersion.map("-RC" + _).getOrElse("")
+      s"${major.value}.${minorVersion}.$patchVersion$maybeRcSuffix-bin-${dateVersion.format(yyyyMMdd)}-$shortSha1-NIGHTLY"
+    // Scala 2 nightly
+    case Nightly(major, minorVersion, patchVersion, _, _, shortSha1) =>
       s"${major.value}.${minorVersion}.$patchVersion-bin-$shortSha1"
   }
 }
@@ -92,6 +107,8 @@ object ScalaVersion {
       major: MajorVersion,
       minorVersion: Int,
       patchVersion: Int,
+      rcVersion: Option[Int],
+      dateVersion: Option[LocalDate],
       shortSha1Version: String
   ) extends ScalaVersion {
     override val minor: Some[Int] = Some(minorVersion)
@@ -126,6 +143,7 @@ object ScalaVersion {
   }
 
   private val intPattern = """\d{1,2}"""
+  private val datePattern = """\d{8}"""
   private val shortSha1Pattern = """[0-9a-f]{7}"""
 
   private val FullVersion =
@@ -133,15 +151,26 @@ object ScalaVersion {
   private val RcVersion =
     raw"""($intPattern)\.($intPattern)\.($intPattern)-RC($intPattern)""".r
   private val NightlyVersion =
-    raw"""($intPattern)\.($intPattern)\.($intPattern)-bin-($shortSha1Pattern)""".r
+    raw"""($intPattern)\.($intPattern)\.($intPattern)(?:-RC($intPattern))?-bin(?:-($datePattern))?-($shortSha1Pattern)(?:-NIGHTLY)?""".r
   private val MajorPattern = raw"""($intPattern)""".r
   private val PartialVersion = raw"""($intPattern)\.($intPattern)""".r
 
+  private val yyyyMMdd = DateTimeFormatter.ofPattern("yyyyMMdd")
+
   def from(version: String): Try[ScalaVersion] = {
     version match {
-      case NightlyVersion(major, minor, patch, shortSha1) =>
+      case NightlyVersion(major, minor, patch, rc, date, shortSha1) =>
         MajorVersion.from(major.toLong).flatMap { major =>
-          Success(Nightly(major, minor.toInt, patch.toInt, shortSha1))
+          Success(
+            Nightly(
+              major,
+              minor.toInt,
+              patch.toInt,
+              Try(rc.toInt).toOption,
+              Try(LocalDate.parse(date, yyyyMMdd)).toOption,
+              shortSha1
+            )
+          )
         }
       case RcVersion(major, minor, patch, rc) =>
         MajorVersion.from(major.toLong).flatMap { major =>

--- a/scalafix-tests/unit/src/test/scala/scalafix/tests/config/ScalaVersionSuite.scala
+++ b/scalafix-tests/unit/src/test/scala/scalafix/tests/config/ScalaVersionSuite.scala
@@ -22,12 +22,28 @@ class ScalaVersionSuite extends munit.FunSuite {
 
   test("parse: 2.13.16-bin-ce78754") {
     val scala2 = ScalaVersion.from("2.13.16-bin-ce78754")
-    assert(scala2.get == Nightly(Major.Scala2, 13, 16, "ce78754"))
+    assert(scala2.get == Nightly(Major.Scala2, 13, 16, None, None, "ce78754"))
+    assert(scala2.get.value == "2.13.16-bin-ce78754")
   }
 
   test("parse: 3.0.0-RC3") {
     val scala3 = ScalaVersion.from("3.0.0-RC3")
     assert(scala3.get == RC(Major.Scala3, 0, 0, 3))
+  }
+
+  test("parse: 3.6.0-RC1-bin-20241008-3408ed7-NIGHTLY") {
+    val scala3 = ScalaVersion.from("3.6.0-RC1-bin-20241008-3408ed7-NIGHTLY")
+    assert(
+      scala3.get == Nightly(
+        Major.Scala3,
+        6,
+        0,
+        Some(1),
+        Some(java.time.LocalDate.parse("2024-10-08")),
+        "3408ed7"
+      )
+    )
+    assert(scala3.get.value == "3.6.0-RC1-bin-20241008-3408ed7-NIGHTLY")
   }
 
   test("parse failure: 3.0.0RC3") {

